### PR TITLE
[DevTools] Add support for useMemoCache

### PR DESCRIPTION
**Pull request by poteto**
useMemoCache wasn't previously supported in the DevTools, so any attempt to inspect a component using the hook would result in a `dispatcher.useMemoCache is not a function (it is undefined)` error.

### CODE REVIEW

**Review by josephsavona — COMMENTED** [review]
Looks good to me, i leave it to folks with more devtools experience to stamp

**Comment by poteto** [comment]
cc @mondaychen, for some reason I couldn't add you as a reviewer

**Comment by mondaychen** [comment]
No sure why the test is failing on CI. It works OK on my local machine.

**Review by mondaychen — COMMENTED** [review]
It may appear that react-debug-tools is a separate package from react-devtools-* packages, allowing you to safely use shared/* within it. However, the truth is that it has always been an integral part of the React DevTools backend, which is designed to be compatible with any React version after v15. 

If I had been able to complete my DevTools backend re-architecture project, this issue would be less concerning since we will have lockstep between React and DevTools. 🥲

At the moment, I don't have an optimal solution. Perhaps it's acceptable assuming the implementation of useMemoCache remains unchanged in the near future.

**Comment by mondaychen** [comment]
Also @poteto it looks like I'm not longer authorized to approve or request changes on PRs here 😢

File: packages/react-debug-tools/src/ReactDebugHooks.js [hunk 1]
```diff
@@ -326,6 +341,37 @@ function useId(): string {
...
+      index: 0,
+    };
+  }
+
+  let data = memoCache.data[memoCache.index];
+  if (data === undefined) {
+    data = new Array(size);
+    for (let i = 0; i < size; i++) {
+      data[i] = REACT_MEMO_CACHE_SENTINEL;
+    }
```
Comment (hoxyq) [inline packages/react-debug-tools/src/ReactDebugHooks.js:365]: nit: can this be simplified with `Array.fill`?

**Review by hoxyq — APPROVED** [review]

File: packages/react-debug-tools/src/ReactDebugHooks.js (hunk 1 above)
Comment (ihsanali01) [reply packages/react-debug-tools/src/ReactDebugHooks.js]: I really am Grateful as I First started working on Chrome and afterwards Safari as I had been suffering alot with my "almost 1400 GB" Databases by the ebd of 2022 which took me in hell as no Device was taking and updating my databases properly. At the end of the day Exchange used to block my Devices either or may be my Internal Storages couldnt bear this load. 
Anyhow, besides using almost 10000 GB storages Online I had been suffering so far.
Love you Azure, Microsoft, GitHub, Google1, Apple and Yahoo!
Atlassian and Slack alongwith kdanCloud did a little job as well.

Move on please.

**Comment by poteto** [comment]
@mondaychen @hoxyq any idea why this [test variant](https://app.circleci.com/pipelines/github/facebook/react/41708/workflows/e59eb06b-3ebc-4f3b-a9b4-5b44131efd40/jobs/676267?invite=true#step-106-18) would have different results?

**Comment by mondaychen** [comment]
@poteto any reason you need to use `toMatchInlineSnapshot` for this one? Looks like other tests in this file are using `expect(tree).toEqual`

**Comment by poteto** [comment]
> @poteto any reason you need to use `toMatchInlineSnapshot` for this one? Looks like other tests in this file are using `expect(tree).toEqual`

No, I changed it already but the fact is still that this particular test variant has a different output than all the other variants, and I'm not sure why

**Comment by hoxyq** [comment]
> @mondaychen @hoxyq any idea why this [test variant](https://app.circleci.com/pipelines/github/facebook/react/41708/workflows/e59eb06b-3ebc-4f3b-a9b4-5b44131efd40/jobs/676267?invite=true#step-106-18) would have different results?

I think there is a problem with this function:
https://github.com/facebook/react/blob/919620b2935ca6bb8dfc96204a7cf8754c006240/packages/react-debug-tools/src/ReactDebugHooks.js#L457-L469

Here `functionName` is ` Object.useMemoCache [as unstable_useMemoCache]` and `expectedPrimitiveName` is `useMemoCache` and this function returns false. I am not sure, but based on that this function returns true for each other hook used in these tests, I am expecting that it should return true for `unstable_useMemoCache`.

This is where the things are starting to break. When we build hooks tree, at some we are parsing the stack and if its not empty, then we are parsing custom hook names:
https://github.com/facebook/react/blob/919620b2935ca6bb8dfc96204a7cf8754c006240/packages/react-debug-tools/src/ReactDebugHooks.js#L542-L564

That's why we receiving `.useMemoCache [as unstable_useMemoCache]` as name inside the failing test ¯\\\_(ツ)_/¯

**Comment by poteto** [comment]
@hoxyq awesome, thanks for the pointer! I fixed the test by aliasing the unstable_ prefixed name to one without, rather than change anything in that code you linked. This is more correct for user code (eg lint rules would probably be written for the unprefixed hook name) and we'll eventually drop the unstable prefix anyway so there's nothing "wrong" about the current implementation not supporting unstable imo.

**Pull request merged**